Keep tooltip open when it gets hovered

Andarist opened the pull request:
Closes #1131 

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [x] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [x] Updated documentation (if it already exists for this component).
- [x] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

`yarn test`

Andarist commented on packages/@react-aria/tooltip/test/useTooltip.test.js:
```diff
@@ -10,18 +10,211 @@
+    jest.useFakeTimers('modern');
+
+    function Test() {
+      useInteractionModality();
```
Having to do that (or using `useFocusVisible` which also leads to `setupGlobalFocusEvents`) is IMHO a bug - but it's a separate issue so I think it would be best to create a ticket for it and handle it later.

Andarist replied:
Filed an issue here: https://github.com/adobe/react-spectrum/issues/1378

Andarist commented on packages/@react-aria/tooltip/test/useTooltip.test.js:
```diff
@@ -10,18 +10,211 @@
+    const triggerElement = container.firstChild.firstChild;
+
+    // trigger pointer modality
+    fireEvent.mouseMove(container);
```
OTOH having to do this sounds OK - I've scratched my head a little bit before I've realized that I have to do that (because `mouseenter` is not handled within `setupGlobalFocusEvents`) but in the real apps `mousemove` will always occur before `mouseenter` on the trigger element so I don't think that changing the implementation to cover testing quirks is required.

snowystinger replied:
yeah, might be good to have a util for this eventually, but we don't need to do it here

Andarist commented on packages/@react-stately/tooltip/src/useTooltipTriggerState.ts:
```diff
@@ -41,10 +41,11 @@ let globalCooldownTimeout = null;
  * the delay for showing a tooltip.
  */
-export function useTooltipTriggerState(props: TooltipTriggerProps): TooltipTriggerState {
+export function useTooltipTriggerState(props: TooltipTriggerProps = {}): TooltipTriggerState {
```
all options are optional so IMHO it makes sense to make `props` optional as well

Andarist commented on packages/@react-stately/tooltip/src/useTooltipTriggerState.ts:
```diff
@@ -41,10 +41,11 @@ let globalCooldownTimeout = null;
   let {delay = TOOLTIP_DELAY} = props;
   let {isOpen, open, close} = useOverlayTriggerState(props);
   let id = useMemo(() => `${++tooltipId}`, []);
+  let closeTimeout = useRef<ReturnType<typeof setTimeout>>();
```
this looks quirky but the return type of `setTimeout` is depending on env and on types loaded for that env so this ensures that it will always work~, regardless of what you are going to put into `tsconfig.json#compilerOptions.lib`

Andarist commented on packages/@react-stately/tooltip/src/useTooltipTriggerState.ts:
```diff
@@ -75,7 +78,13 @@ export function useTooltipTriggerState(props: TooltipTriggerProps): TooltipTrigg
+      closeTimeout.current = setTimeout(() => {
+        closeTimeout.current = null
+        close()
+      }, TOOLTIP_COOLDOWN)
```
I've chosen to reuse `TOOLTIP_COOLDOWN` which is equal to 500 but it might be worth shortening this timeout here, thoughts?

snowystinger replied:
I think we can probably make it a prop to pass to useTooltipTriggerState if other libraries need to control that, but we can start with this default. For the record, it's not specified in Spectrum either right now. https://spectrum.adobe.com/page/tooltip/#Behaviors

Andarist commented on packages/@react-aria/tooltip/src/useTooltip.ts:
```diff
@@ -24,11 +26,16 @@ interface TooltipAria {
  * Provides the accessibility implementation for a Tooltip component.
  */
-export function useTooltip(props: AriaTooltipProps): TooltipAria {
+export function useTooltip(props: AriaTooltipProps, state?: TooltipTriggerState): TooltipAria {
```
I've chosen to accept `TooltipTriggerState` here because it contains everything that we need here. I couldn't make it required right now as it would be a breaking change but I believe that this should be done in the future.

devongovett replied:
This is fine I think. Maybe we should add a `console.warn` here in case `state` isn't passed? That way it's not technically breaking but people know to update their code.

Andarist replied:
Yeah - this could be done. I was wondering - any particular reason why you have not been using `process.env.NODE_ENV` checks in the codebase so far? Such a warning would be a great candidate for a dev-only branch of code.

Andarist commented on packages/@react-spectrum/tooltip/test/TooltipTrigger.test.js:
```diff
@@ -469,7 +486,7 @@ describe('TooltipTrigger', function () {
     });
 
-    it('once opened, it can be closed and opened instantly for a period of time',  () => {
+    it.skip('once opened, it can be closed and opened instantly for a period of time',  () => {
```
this test doesn't make a lot of sense now unless you'd like to specify different global and local cooldowns

snowystinger replied:
Maybe we change this test to use the `ESC` key to close it faster so we can check the cooldown

Andarist replied:
good idea - I've just refactored this test using the advice 👍

dannify commented:
Thanks for the contribution @Andarist! We will take a look.

snowystinger reviewed (commented):
Hey! Thanks for starting this, I see the delay that's been added to closing. However, it doesn't look like the state has been passed through in the aria docs nor the react spectrum component, so it doesn't actually stay open on hover of the tooltip. Is there a missing commit? or was this the intent?

Andarist commented:
@snowystinger I've focused on implementing `@react-aria` part as `@react-spectrum` is out of the scope of my interests 😉 It wasn't really too complicated so I've added this just now - please take a look. As to the docs - I've also added this to the existing example, not sure if this is enough or you'd like to add more information about this.

snowystinger commented:
@Andarist That's understandable, I just wanted to make sure that was your intent. I appreciate that you added it to the other places, makes it easier to play with a live example.

snowystinger commented on packages/@react-aria/tooltip/docs/useTooltipTrigger.mdx:
```diff
@@ -112,7 +112,7 @@ function Example(props) {
   return (
     <span style={{position: 'relative'}}>
-      <button ref={ref} {...triggerProps}>I have a tooltip</button>
+      <button ref={ref} state={state} {...triggerProps}>I have a tooltip</button>
```
I think you meant to add this to `Tooltip`?

Andarist replied:
yeah, this was done in a hurry 🤦 fixed

snowystinger commented on packages/@react-aria/tooltip/test/useTooltip.test.js:
```diff
@@ -10,18 +10,211 @@
+
+      const state = useTooltipTriggerState({delay: 0});
+      const {triggerProps, tooltipProps} = useTooltipTrigger({}, state);
+      const {tooltipProps: finalTooltipProps} = useTooltip(tooltipProps);
```
we should pass `state` to `useTooltip` given your comment about this should be required? same comment for the other instances in this file
or is this to make sure that it works both ways?

Andarist replied:
> or is this to make sure that it works both ways?

This. As mentioned in some other comment - I think that this should be made required in the future but that would be a breaking change. We could add a single test ensuring that though and just make the rest use `state` - I'm fine either way, just let me know what do you think is best here.

snowystinger replied:
I think we can leave as is then, I just wanted to make sure it was intentional.

snowystinger commented on packages/@react-aria/tooltip/test/useTooltip.test.js:
```diff
@@ -10,18 +10,211 @@
+
+      return (
+        <div>
+          <span {...triggerProps}>{'Trigger'}</span>
```
```suggestion
          <span {...triggerProps}>Trigger</span>
```

Andarist replied:
Changed all occurrences of this pattern in a file 👍

snowystinger commented on packages/@react-aria/tooltip/test/useTooltip.test.js:
```diff
@@ -10,18 +10,211 @@
+      return (
+        <div>
+          <span {...triggerProps}>{'Trigger'}</span>
+          {state.isOpen && <span {...finalTooltipProps}>{'Tooltip'}</span>}
```
```suggestion
          {state.isOpen && <span {...finalTooltipProps}>Tooltip</span>}
```

Andarist commented on packages/@react-aria/tooltip/test/useTooltip.test.js:
```diff
@@ -10,18 +10,211 @@
+    }
+
+    const {container, queryByRole} = render(<Test />);
+    const triggerElement = container.firstChild.firstChild;
```
Good call - I've refactored this to use a button as a trigger element and this makes it easier to query it by role.

snowystinger reviewed (approved):
LGTM, I'll let the team know it's ready for its second review

ktabors reviewed (approved):
LGTM

Andarist commented:
One of the last thoughts on this one - if the same amount of time is used for both keeping the tooltip open after leaving the trigger and for the global "cooldown" then how useful the cooldown mechanism is? It still makes a functional difference for situations when a tooltip gets closed with ESC as noticed by @snowystinger but as the whole cooldown mechanism is focused on pointer interactions I'm not sure if this situation is really important - I would be inclined to say that no, it isn't. If you agree with that then the implementation could be simplified slightly by dropping that mechanism and making the warmup to rely on the module state of opened tooltips.

snowystinger commented:
@Andarist Cooldown is different from time allowed to get hovered over the tooltip itselt. Cooldown is global to all tooltips, so if you hover one, it pops up, now if you hover a different target, the tooltip will appear instantly. There's a series of stories with multiple tooltips where you can try this out if you like, it's also explained in our documentation here https://react-spectrum.adobe.com/react-spectrum/Tooltip.html#warmup--cooldown

Andarist commented:
I understand this mechanism - I might have not described my intention too well though so let me rephrase.

I believe that the cooldown could be removed because when leaving a trigger a tooltip doesn't disappear immediately (unlike before the changes from this PR). The disappearing of the tooltip could act as a cooldown mechanism - if we hover another trigger element when the previous tooltip is still displayed then we switch to displaying the new tooltip anyway (as the mechanism has been warmed up already) and if the tooltip disappears then we could consider this as cooldown delay passed so if we hover over whatever trigger after that then we'll have to wait for a new warmup.

snowystinger commented:
Ahhhh, yes, I misunderstood. Thank you for clarifying. I'll ask Design, though we may not get an answer back until the new year.

Andarist commented:
Sure, no problem. This doesn't have to block this PR though - it's just a potential refactor opportunity for the future. Do you think that those 500ms is the appropriate time for keeping the tooltip open when leaving the trigger?

snowystinger commented:
Yep, 500ms is fine for this PR
